tests/test_request/test_request.py

```diff
@@ -0,0 +1,276 @@
+    def test_submit_should_assign_worker_and_execute(self):
+        stop = threading.Event()
+
+        def work():
+            time.sleep(0.3)
+            stop.set()
+            return 42
+
+        req = Request(work)
+        req.submit()
+        assert req.wait() == 42
+        assert req.assigned_worker in Request.global_thread_pool.workers
+
+    def test_submit_dependent_requests_should_execute_on_same_worker(self):
+        stop = threading.Event()
+
+        more_work = Request(Work(lambda: 42))
+
+        req = Request(Work(lambda: more_work.wait()))
+        req.submit()
+
+        assert req.wait() == 42
+        assert req.assigned_worker in Request.global_thread_pool.workers
+        assert req.assigned_worker == more_work.assigned_worker
+
+    @pytest.fixture
+    def broken_fn(self, request):
+        def broken():
+            raise TExc()
+
+        return broken
+
+    @pytest.fixture
+    def work_fn(self, request):
+        def work_fn():
+            return 42
+        return work_fn
+
+    @staticmethod
+    def work_req(fn):
+        work = Work(fn)
+        req = Request(work)
+        return work, req
+
+    def test_signal_finished_called_on_completion(self, work_fn):
+        work, req = self.work_req(work_fn)
+
+        recv = mock.Mock()
+
+        req.notify_finished(recv)
+        req.submit()
+        assert work.done.wait(timeout=1)
+
+        recv.assert_called_once_with(42)
+
+    def test_signal_finished_called_when_subscription_happened_after_completion(self, work_fn):
+        work, req = self.work_req(work_fn)
+
+        recv = mock.Mock()
+        req.submit()
+        assert work.done.wait(timeout=1)
+
+        req.notify_finished(recv)
+
+        recv.assert_called_once_with(42)
+
+    def test_signal_finished_should_not_be_called_on_exception(self, broken_fn):
+        work, req = self.work_req(broken_fn)
+        recv = mock.Mock()
+
+        req.notify_finished(recv)
+        req.submit()
+
+        with pytest.raises(TExc):
+            assert req.wait() == 42
+        recv.assert_not_called()
+
+    def test_signal_failed_should_be_called_on_exception(self, broken_fn):
+        work, req = self.work_req(broken_fn)
+
+        recv = mock.Mock()
+
+        req = Request(work)
+        req.notify_failed(recv)
+        req.submit()
+
+        with pytest.raises(TExc):
+            assert req.wait() == 42
+        recv.assert_called_once()
+        assert isinstance(recv.call_args[0][0], TExc)
+
+    def test_signal_failed_called_even_when_subscription_happened_after_completion(self, broken_fn):
+        work, req = self.work_req(broken_fn)
+
+        recv = mock.Mock()
+
+        req = Request(work)
+        req.submit()
+
+        with pytest.raises(TExc):
+            assert req.wait() == 42
+
+        req.notify_failed(recv)
+        recv.assert_called_once()
+        assert isinstance(recv.call_args[0][0], TExc)
+
+
+@pytest.fixture
+def work():
+    cont = threading.Event()
```

Comment:
could be at least something like `continue` or `unpause`
```suggestion
    continue = threading.Event()
```